src/OperationProcessor.ts

```diff
@@ -76,576 +18,145 @@ export interface OperationProcessor {
    * @param did The DID to resolve. e.g. did:sidetree:abc123.
    * @returns DID Document of the given DID. Undefined if the DID is deleted or not found.
+   *
+   * Iterate over all operations in blockchain-time order extending the
+   * the operation chain while checking validity.
    */
   public async resolve (did: string): Promise<DidDocument | undefined> {
-    const didUniqueSuffix = did.substring(this.didMethodName.length);
-
-    await this.processDeferredOperationsOfDid(didUniqueSuffix);
+    let didDocument: DidDocument | undefined;
+    let previousOperation: Operation | undefined;
 
-    if (this.deletedDids.has(did)) {
-      return undefined;
-    }
-
-    const latestVersion = await this.last(didUniqueSuffix);
+    const didUniqueSuffix = did.substring(this.didMethodName.length);
 
-    // lastVersion === undefined implies we do not know about the did
-    if (latestVersion === undefined) {
-      return undefined;
+    const didOps = await this.operationStore.get(didUniqueSuffix);
+    for (const operation of didOps) {
+      if (await this.isValid(operation, previousOperation, didDocument)) {
+        didDocument = this.getUpdatedDocument(didDocument, operation);
+        previousOperation = operation;
+      }
     }
 
-    return this.lookup(latestVersion);
+    return didDocument;
   }
 
   /**
-   * Returns the DID Document for a given version identifier.
+   * Is an operation valid in the context of a resolve() when extending
+   * operation chain for a DID. When this function is called, resolve() has
+   * constructed a valid operation that starts with some Create operation and
+   * ends with previousOperation, resulting the DID Document 'currentDidDocument'.
+   * This function checks if parameter operation can be used to legitimately
+   * extend this chain.
    */
-  public async lookup (versionId: VersionId): Promise<DidDocument | undefined> {
-    // Version id is also the operation hash that produces the document
-    const opHash = versionId;
-
-    const opInfo = this.opHashToInfo.get(opHash);
+  private async isValid (operation: Operation, previousOperation: Operation | undefined, currentDidDocument: DidDocument | undefined):
+    Promise<boolean> {
 
-    // We don't know anything about this operation
-    if (opInfo === undefined) {
-      return undefined;
-    }
+    if (operation.type === OperationType.Create) {
 
-    // Construct the operation using a CAS lookup
-    const op = await this.operationStore.lookup(opHash);
+      // If either of these is defined, then we have seen a previous create operation.
+      if (previousOperation || currentDidDocument) {
+        return false;
+      }
 
-    if (this.isInitialVersion(opInfo)) {
-      const protocolVersion = Protocol.getProtocol(op.transactionTime!);
-      return Document.from(op.encodedPayload, this.didMethodName, protocolVersion.hashAlgorithmInMultihashCode);
+      // TODO: Add signature verification for create operation
```

Comment:
Need to add the logic else the regression test I added in master will fail.

After rebase, you should be able to use `Document.from(...)` and `operation.verifySignature(...) to do so.

Reply:
Added.